docs/source/notes/ddp.rst

```diff
@@ -0,0 +1,191 @@
+.. _ddp:
+
+Distributed Data Parallel
+=========================
+
+.. warning::
+  The implementation of :class:`torch.nn.parallel.DistributedDataParallel`
+  evolves over time. This design note is written based on the states as of v1.4.
+
+
+:class:`torch.nn.parallel.DistributedDataParallel` (DDP) replicates model,
+splits input data, and transparently performs distributed data parallel
+training. This page describes how it works and reveals implementation details.
```

Comment:
Not entirely true, because it doesn't do input data splitting (done by the `DistributedSampler`). Also, it doesn't automatically replicate across machines (and I think we shouldn't promote usage of the built-in replication). I think saying "transparently performs distributed data parallel training of an arbitrary `nn.Module`" is enough here.